File: native/android/core-model/src/main/kotlin/com/fireball/nativeapp/core/model/PlaybackAdvanceRules.kt
package com.fireball.nativeapp.core.model

/**
 * Shared queue advance decisions for native players (parity with iOS [PlaybackAdvanceRules]).
 */
enum class PlaybackRepeatMode {
    OFF,
    ALL,
    ONE,
    ;

    companion object {
        fun fromPersisted(value: String): PlaybackRepeatMode =
            when (value.trim().uppercase()) {
                "ALL" -> ALL
                "ONE" -> ONE
                else -> OFF
            }
    }
}

sealed class PlaybackAdvanceDecision {
    data object RestartCurrent : PlaybackAdvanceDecision()

    data class JumpToIndex(val index: Int) : PlaybackAdvanceDecision()

    data object StopPlayback : PlaybackAdvanceDecision()

    /** Next pressed at tail with no wrap (no-op). */
    data object NoOp : PlaybackAdvanceDecision()
}

object PlaybackAdvanceRules {
    fun afterTrackFinished(
        currentIndex: Int,
        queueSize: Int,
        repeatMode: PlaybackRepeatMode,
        shuffled: Boolean,
        queueModeRepeat: Boolean,
        pickShuffleIndex: (List<Int>) -> Int?,
    ): PlaybackAdvanceDecision {
        if (currentIndex < 0 || queueSize <= 0) return PlaybackAdvanceDecision.NoOp
        if (repeatMode == PlaybackRepeatMode.ONE) return PlaybackAdvanceDecision.RestartCurrent
        if (shuffled && queueSize > 1) {
            val choices = (0 until queueSize).filter { it != currentIndex }
            val next = pickShuffleIndex(choices)
            return if (next != null) PlaybackAdvanceDecision.JumpToIndex(next) else PlaybackAdvanceDecision.NoOp
        }
        if (currentIndex + 1 < queueSize) return PlaybackAdvanceDecision.JumpToIndex(currentIndex + 1)
        if (repeatMode == PlaybackRepeatMode.ALL) return PlaybackAdvanceDecision.JumpToIndex(0)
        if (queueModeRepeat) return PlaybackAdvanceDecision.JumpToIndex(0)
        return PlaybackAdvanceDecision.StopPlayback
    }

    fun userPressedNext(
        currentIndex: Int,
        queueSize: Int,
        repeatMode: PlaybackRepeatMode,
        shuffled: Boolean,
        queueModeRepeat: Boolean,
        pickShuffleIndex: (List<Int>) -> Int?,
    ): PlaybackAdvanceDecision {
        if (currentIndex < 0 || queueSize <= 0) return PlaybackAdvanceDecision.NoOp
        if (repeatMode == PlaybackRepeatMode.ONE) return PlaybackAdvanceDecision.RestartCurrent
        if (shuffled && queueSize > 1) {
            val choices = (0 until queueSize).filter { it != currentIndex }
            val next = pickShuffleIndex(choices)
            return if (next != null) PlaybackAdvanceDecision.JumpToIndex(next) else PlaybackAdvanceDecision.NoOp
        }
        if (currentIndex + 1 < queueSize) return PlaybackAdvanceDecision.JumpToIndex(currentIndex + 1)
        if (repeatMode == PlaybackRepeatMode.ALL) return PlaybackAdvanceDecision.JumpToIndex(0)
        if (queueModeRepeat) return PlaybackAdvanceDecision.JumpToIndex(0)
        return PlaybackAdvanceDecision.NoOp
    }

    fun userPressedPrevious(
        currentIndex: Int,
        positionSeconds: Double,
        repeatMode: PlaybackRepeatMode,
        shuffled: Boolean,
        queueSize: Int,
        pickShuffleIndex: (List<Int>) -> Int?,
    ): PlaybackAdvanceDecision {
        if (currentIndex < 0 || queueSize <= 0) return PlaybackAdvanceDecision.NoOp
        if (repeatMode == PlaybackRepeatMode.ONE) return PlaybackAdvanceDecision.RestartCurrent
        if (positionSeconds > 3.0) return PlaybackAdvanceDecision.RestartCurrent
        if (shuffled && queueSize > 1) {
            val choices = (0 until queueSize).filter { it != currentIndex }
            val pick = pickShuffleIndex(choices)
            return if (pick != null) PlaybackAdvanceDecision.JumpToIndex(pick) else PlaybackAdvanceDecision.NoOp
        }
        return PlaybackAdvanceDecision.JumpToIndex((currentIndex - 1).coerceAtLeast(0))
    }
}

File: native/ios/FireballNative/Core/PlaybackAdvanceRules.swift
import Foundation

/// Shared queue advance decisions for native players (parity with Android `PlaybackAdvanceRules`).
enum PlaybackRepeatMode: Equatable {
    case off
    case all
    case one
}

enum PlaybackAdvanceDecision: Equatable {
    case restartCurrent
    case jumpTo(Int)
    case stopPlayback
    case noOp
}

enum PlaybackAdvanceRules {
    static func afterTrackFinished(
        currentIndex: Int,
        queueSize: Int,
        repeatMode: PlaybackRepeatMode,
        shuffled: Bool,
        queueModeRepeat: Bool,
        pickShuffleIndex: ([Int]) -> Int?
    ) -> PlaybackAdvanceDecision {
        guard currentIndex >= 0, queueSize > 0 else { return .noOp }
        if repeatMode == .one { return .restartCurrent }
        if shuffled, queueSize > 1 {
            let choices = (0..<queueSize).filter { $0 != currentIndex }
            if let next = pickShuffleIndex(choices) { return .jumpTo(next) }
            return .noOp
        }
        if currentIndex + 1 < queueSize { return .jumpTo(currentIndex + 1) }
        if repeatMode == .all { return .jumpTo(0) }
        if queueModeRepeat { return .jumpTo(0) }
        return .stopPlayback
    }

    static func userPressedNext(
        currentIndex: Int,
        queueSize: Int,
        repeatMode: PlaybackRepeatMode,
        shuffled: Bool,
        queueModeRepeat: Bool,
        pickShuffleIndex: ([Int]) -> Int?
    ) -> PlaybackAdvanceDecision {
        guard currentIndex >= 0, queueSize > 0 else { return .noOp }
        if repeatMode == .one { return .restartCurrent }
        if shuffled, queueSize > 1 {
            let choices = (0..<queueSize).filter { $0 != currentIndex }
            if let next = pickShuffleIndex(choices) { return .jumpTo(next) }
            return .noOp
        }
        if currentIndex + 1 < queueSize { return .jumpTo(currentIndex + 1) }
        if repeatMode == .all { return .jumpTo(0) }
        if queueModeRepeat { return .jumpTo(0) }
        return .noOp
    }

    static func userPressedPrevious(
        currentIndex: Int,
        positionSeconds: Double,
        repeatMode: PlaybackRepeatMode,
        shuffled: Bool,
        queueSize: Int,
        pickShuffleIndex: ([Int]) -> Int?
    ) -> PlaybackAdvanceDecision {
        guard currentIndex >= 0, queueSize > 0 else { return .noOp }
        if repeatMode == .one { return .restartCurrent }
        if positionSeconds > 3 { return .restartCurrent }
        if shuffled, queueSize > 1 {
            let choices = (0..<queueSize).filter { $0 != currentIndex }
            if let pick = pickShuffleIndex(choices) { return .jumpTo(pick) }
            return .noOp
        }
        return .jumpTo(max(0, currentIndex - 1))
    }
}
